drivers/adc/adc_max11102_17.c

```diff
@@ -0,0 +1,454 @@
+
+	k_sem_init(&data->acquire_signal, 0, 1);
+
+	if (!spi_is_ready_dt(&config->bus)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	switch (config->channel_count) {
+	case 1:
+		if (config->gpio_chsel.port != NULL) {
+			LOG_ERR("GPIO for chsel set with only one channel");
+			return -EINVAL;
+		}
+		break;
+	case 2:
+		if (config->gpio_chsel.port == NULL) {
+			LOG_ERR("no GPIO for chsel set with two channels");
+			return -EINVAL;
+		}
+
+		result = gpio_pin_configure_dt(&config->gpio_chsel, GPIO_OUTPUT_INACTIVE);
+		if (result != 0) {
+			LOG_ERR("failed to initialize GPIO for chsel");
+			return result;
+		}
+		break;
+	default:
+		LOG_ERR("invalid number of channels (%i)", config->channel_count);
+		return -EINVAL;
+	}
+
+	data->current_channel_id = 0;
+
+#if CONFIG_ADC_ASYNC
+	const k_tid_t tid = k_thread_create(
+		&data->thread, data->stack, CONFIG_ADC_MAX11102_17_ACQUISITION_THREAD_STACK_SIZE,
+		(k_thread_entry_t)max11102_17_acquisition_thread, (void *)dev, NULL, NULL,
+		CONFIG_ADC_MAX11102_17_ASYNC_THREAD_INIT_PRIO, 0, K_NO_WAIT);
+	k_thread_name_set(tid, "adc_max11102_17");
+#endif
+
+	/* power up time is one conversion cycle */
+	result = max11102_17_read_sample(dev, &sample);
+	if (result != 0) {
+		LOG_ERR("unable to read dummy sample for power up timing");
+		return result;
+	}
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return result;
+}
+
+static const struct adc_driver_api api = {
+	.channel_setup = max11102_17_channel_setup,
+	.read = max11102_17_read,
+	.ref_internal = 0,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = max11102_17_adc_read_async,
+#endif
+};
+
+BUILD_ASSERT(CONFIG_ADC_INIT_PRIORITY > CONFIG_SPI_INIT_PRIORITY,
+	     "CONFIG_ADC_INIT_PRIORITY must be higher than CONFIG_SPI_INIT_PRIORITY");
+
+#define ADC_MAX11102_17_INST_DEFINE(index, name, res, channels)                                    \
+	static const struct max11102_17_config config_##name##_##index = {                         \
+		.bus = SPI_DT_SPEC_INST_GET(                                                       \
+			index,                                                                     \
+			SPI_OP_MODE_MASTER | SPI_MODE_CPOL | SPI_MODE_CPHA | SPI_WORD_SET(8), 0),  \
+		.gpio_chsel = GPIO_DT_SPEC_INST_GET_OR(index, chsel_gpios, {0}),                   \
+		.resolution = res,                                                                 \
+		.channel_count = channels,                                                         \
+	};                                                                                         \
+	static struct max11102_17_data data_##name##_##index;                                      \
+	DEVICE_DT_INST_DEFINE(index, max11102_17_init, NULL, &data_##name##_##index,               \
+			      &config_##name##_##index, POST_KERNEL, CONFIG_ADC_INIT_PRIORITY,     \
+			      &api);
+
+#define DT_DRV_COMPAT           maxim_max11102
+#define ADC_MAX11102_RESOLUTION 12
+#define ADC_MAX11102_CHANNELS   2
```

Comment:
Do you actually need to define those?
Wouldn't something like this work:
```
/* 12-bit, 2 channels */
DT_INST_FOREACH_STATUS_OKAY_VARGS(ADC_MAX11102_17_INST_DEFINE, 12, 2)
```
with `ADC_MAX11102_17_INST_DEFINE` defined as:
```
#define ADC_MAX11102_17_INST_DEFINE(index, res, channels) \
	_ADC_MAX11102_17_INST_DEFINE(index, DT_DRV_COMPAT, res, channels)
```
and `_ADC_MAX11102_17_INST_DEFINE` as the current `ADC_MAX11102_17_INST_DEFINE`?

Reply:
Yes, might work. But then you have the information duplicated (comment and actual macro parameter). I also assume having these defines within the c-file should not create any name clashes and they increase the readability in my opinion.